src/Core/multiplier.scala

```diff
@@ -17,14 +17,60 @@ class Multiplier(implicit conf: FlexpretConfiguration) extends Module {
     val func   = UInt(INPUT, 4)
     val result = Bits(OUTPUT, 32)
   }
-  
-  val op1 = Mux(io.func === MUL_HU, Cat(Bits(0, 1), io.op1).toSInt, Cat(io.op1(31), io.op1).toSInt)
-  val op2 = Mux(io.func === MUL_HSU || io.func === MUL_HU, Cat(Bits(0, 1), io.op2).toSInt, Cat(io.op2(31), io.op2).toSInt)
-  val mul_result = op1 * op2
-  val result = Mux(io.func === MUL_L, mul_result(31, 0), mul_result(63, 32))
+
+  val op1 = Mux(io.func === MUL_HU || io.func === DIV_LU || io.func === REM_LU, Cat(Bits(0, 1), io.op1).toSInt, Cat(io.op1(31), io.op1).toSInt)
+  val op2 = Mux(io.func === MUL_HSU || io.func === MUL_HU || io.func === DIV_LU || io.func === REM_LU, Cat(Bits(0, 1), io.op2).toSInt, Cat(io.op2(31), io.op2).toSInt)
+  val mul_result = Mux(io.func === REM_L || io.func === REM_LU, rem(op1, op2),
+                      Mux(io.func === DIV_L || io.func === DIV_LU, div(op1, op2), op1 * op2));
+  val result = Mux(io.func === MUL_L || io.func === DIV_L || io.func === DIV_LU || io.func === REM_L || io.func === REM_LU, mul_result(31, 0), mul_result(63, 32))
 
   // 2 cycle
   io.result := Reg(next = result)
 
-}
+  def rem(dividend: Bits, divider: Bits) : Bits = {
+    return divider_signed(dividend, divider, Bool(true));
+  }
+
+  def div(dividend : Bits, divider: Bits) : Bits = {
+    return divider_signed(dividend, divider, Bool(false));
+  }
+
+  def divider_signed(dividend: Bits, divider: Bits, isRemainder: Bool) : Bits = {
```

Comment:
`op1` and `op2` are already 33 bit signed values, so I if they are passed in as `SInt` instead of `Bits` I would guess you could use `-dividend` instead of manually taking 2s complement. Also, `SInt` may be comparable against `0` for positive or negative instead of checking the top bit.

Reply:
Changed in commit 410f4b9.